esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -5878,55 +5880,77 @@ int CWsDfuEx::GetIndexData(IEspContext &context, bool bSchemaOnly, const char* i
-            int *locationsIndex = locationMap.getValue(host.str());
-            if (locationsIndex)
+            unsigned *locationIndex = partLocationMap.getValue(host.str());
+            if (locationIndex)
+                fileCopy->setLocationIndex(*locationIndex);
+            else
             {
-                partCopyLocationsIndexes.append(locationsIndexStr.append(*locationsIndex).str());
-                continue;
-            }
+                partLocationMap.setValue(host.str(), ++newLocationIndex);
+                fileCopy->setLocationIndex(newLocationIndex);
 
-            Owned<IEspDFUPartLocations> partLocations = createDFUPartLocations();
-            partLocations->setLocationIndex(nextLocationsIndex);
-            partLocations->setLocation(host.str());
-            dfuPartLocations.append(*partLocations.getClear());
+                Owned<IEspDFUPartLocation> partLocation = createDFUPartLocation();
+                partLocation->setLocationIndex(newLocationIndex);
+                partLocation->setHost(host.str());
+                dfuPartLocations.append(*partLocation.getClear());
+            }
 
-            partCopyLocationsIndexes.append(locationsIndexStr.append(nextLocationsIndex).str());
-            locationMap.setValue(host.str(), nextLocationsIndex);
-            nextLocationsIndex++;
+            fileCopies.append(*fileCopy.getClear());
         }
 
-        Owned<IEspDFUPartCopies> partCopies = createDFUPartCopies();
-        partCopies->setPartIndex(partIndex + 1);
-        partCopies->setLocationIndexes(partCopyLocationsIndexes);
-        dfuPartCopies.append(*partCopies.getClear());
+        Owned<IEspDFUFilePart> filePart = createDFUFilePart();
+        filePart->setPartIndex(part.queryPartIndex() + 1);
+        filePart->setCopies(fileCopies);
+        filePart->setIsTLK(isPartTLK(&part));
+        dfuParts.append(*filePart.getClear());
     }
-    return df->numParts();
+
+    accessInfo.setNumParts(numParts);
+    accessInfo.setFileParts(dfuParts);
+    accessInfo.setFileLocations(dfuPartLocations);
 }
 
 static const char *securityInfoVersion="1";
-void CWsDfuEx::getFileMeta(StringBuffer &metaInfoStr, IDistributedFile &file, IUserDescriptor *user, CFileAccessRole role, const char *expiryTime, const char *keyPairName, IConstDFUFileAccessRequest &req)
+void CWsDfuEx::getFileMeta(StringBuffer &metaInfoStr, StringBuffer &expiryTime, const char *fileName,
+    IFileDescriptor *fDesc, IUserDescriptor *user, const char *jobId, IConstDFUFileAccessRequestBase &req)
 {
+    // setup "expiryTime"
+    unsigned expirySecs = req.getExpirySeconds();
+    if (expirySecs > maxFileAccessExpirySeconds)
+        expirySecs = maxFileAccessExpirySeconds;
+    time_t now;
+    time(&now);
+    CDateTime expiryDt;
+    expiryDt.set(now + expirySecs);
+    expiryDt.getString(expiryTime);
+    
+    StringBuffer keyPairName = env->getClusterKeyPairName(req.getCluster());
```

Comment:
minor - not a compelling reason to make a copy here.
Should change to const char *